### ISSUE TITLE
Efficient randomized initializers for Array/Vector/Matrix

### ISSUE THREAD

**Issue by regexident** [Jounce/Surge#144]
## tl;dr

It would be nice if Surge provided APIs for efficiently generating random numbers of different distributions. Especially being able to generate normal distributed random numbers in parallel would be of great for many stochastic algorithms.

## What

Many bayesian/stochastic processes make use of random values.

As such it would be nice to be able to efficiently initialize instances of `[Scalar]`/`Vector<Scalar>`/`Matrix<Scalar>` with **uniform**-distributed and/or **normal**-distributed values.

## Why

While we will —as far as I can see— not be able to provide functions for generating **uniform**-distributed values, that are any more efficient than, say …

```swift
let rnd: [Float] = (0..<count).map { _ in Float.random(in: range, using: &generator) }
```

… the **normal**-distributed case, which commonly is implemented using the Box-Muller transform …

```swift
let rnd: [Float] = (0..<count).map {
    let u1 = Float.random(in: range, using: &generator)
    let u2 = Float.random(in: range, using: &generator)
    return sqrt(-2.0 * log(u1)) .* cos(2.0 * .pi * u2)
}

```

## How

… can be SIMD-optimized using a parallelized version of the transform …

```swift
let rnd: [Float] = {
    let u1 = (0..<count).map { _ in Float.random(in: range, using: &generator) }
    let u2 = (0..<count).map { _ in Float.random(in: range, using: &generator) }
    return sqrt(-2.0 * log(u1)) .* cos(2.0 * .pi * u2)
}()
```

The individual calculations are now performed in parallel. Nice!

But why stop here?

Instead of creating (and shortly later disposing) a temporary array on every partial expression we perform them efficiently in-place, allocating and re-using nothing but the initial buffers `u1` and `u2`:

```swift
let u1 = (0..<count).map { _ in Float.random(in: range, using: &generator) }
let u2 = (0..<count).map { _ in Float.random(in: range, using: &generator) }

logInPlace(&u1) // 0
mulInPlace(&u1, -2.0) // 1
sqrtInPlace(&u1) // 2

mulInPlace(&u2, 2.0 * .pi) // 3
cosInPlace(&u2 // 4

elmulInPlace(&u1, u2) // 5

let rnd: [Float] = u1
```

The in-place operations correspond to above formula like this:

```
┌─2────────────────┐
│    ┌─1──────────┐│
│    │      ┌─0───┐│
sqrt(-2.0 * log(u1)) .* cos(2.0 * .pi * u2)
│                           └─3──────────┘│
│                       └─4───────────────┘
└─5───────────────────────────────────────┘
```

The individual calculations are still performed in parallel, but without any intermediate allocations which otherwise would be discarded shortly after.

(I already have a proof-of-concept for this on a local branch and will push it as a PR at some point.)
